### PR TITLE
Eiffel on Cloud Events

### LINKED ISSUE #399: Eiffel on Cloud Events
### Summary

**As** CI/CD developer
**I want** an explanation of how to use Eiffel on Cloud Events
**So that** I can use other brokers than RabbitMQ

### Context

When Eiffel was created https://cloudevents.io/ did not exist. As it now becomes a standard for sending events Eiffel should explain how to send Eiffel events on top of cloud events. This has two benefits:
- Eiffel becomes cloud event compatible: Even though it might not add any technical benefit it becomes a selling point
- It becomes easier to send Eiffel events on other transport protocols other than AMQP: The Eiffel protocol does not specify any details on how to send the event (Sepia does that) but prepare for other transport protocols we need a description on how to bind Eiffel events to Cloud Events.

### Exemplification

_No response_

### Drawbacks

_No response_

### Out of Scope

Should not consider how to actually send the events or any details on specific binding toward transport protocols but rather how to bind them to Cloud Events.

### Further links

- https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md
- https://github.com/cdevents/spec/blob/main/cloudevents-binding.md

### Acceptance Criteria

_No response_

### Implementation Ideas

_No response_

### PR DESCRIPTION
**Pull request by m-linner-ericsson**
### Applicable Issues

Fixes #399 

### Description of the Change

Adding a description on how to bind Eiffel to Cloud Events

### Alternate Designs

This PR discusses two different options on how to set the type, see text

### Possible Drawbacks

Opening up possibilities for sending Eiffel events on over other transport protocols than AMQP might give some light of not having just one way of doing things.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by:  Mattias Linnér mattias.linner@ericsson.com

### CODE REVIEW

**Comment by e-backmark-ericsson** [comment]
I really like that this idea of binding Eiffel on to CloudEvents is now getting some realization! Thanks @m-linner-ericsson !

Regarding the alternatives on the type field, I believe the reason for CloudEvents strongly recommending a reverse dns name as a prefix is to make the event type unique across different event 'protocols' (different data payload structures). Therefore I'd recommend we go for such a syntax on the type field. We could consider to make it all lower case, potentially prefixing internal capital letters with hyphen or underscore before the corresponding lowercase letter.

Furthermore I believe we should add the optional dataschema field to the cloudevent part, giving it the same content as our meta.schemaUri

**Comment by m-linner-ericsson** [comment]
> I really like that this idea of binding Eiffel on to CloudEvents is now getting some realization! Thanks @m-linner-ericsson !
> 
> Regarding the alternatives on the type field, I believe the reason for CloudEvents strongly recommending a reverse dns name as a prefix is to make the event type unique across different event 'protocols' (different data payload structures). Therefore I'd recommend we go for such a syntax on the type field. We could consider to make it all lower case, potentially prefixing internal capital letters with hyphen or underscore before the corresponding lowercase letter.
> 
> Furthermore I believe we should add the optional dataschema field to the cloudevent part, giving it the same content as our meta.schemaUri

Good points, fixed that in 0f248f6

**Review by e-backmark-ericsson — APPROVED** [review]

File: eiffel-syntax-and-usage/cloud-events.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+We try to satisfy the reverse DNS name and create one
+
+`io.github.eiffel-community.<Category>.<Eiffel type>`
+
+Where:
+
+- `io.github.eiffel-community` is the reverse DNS of our website
+- `<Category>` is taken from [Event Categories](event-categories.md) but lowercase
+- `<Eiffel type>` is the value from `meta.type` replacing capital letter with lowercase and hyphen
```
Comment (magnusbaeck) [inline eiffel-syntax-and-usage/cloud-events.md:97]: I think such transforms should be as simple as possible. Why not just lowercase meta.type? Or use it as-is? The spec doesn't appear to mandate lowercase.

File: eiffel-syntax-and-usage/cloud-events.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+### Alternative 1
+
+We try to satisfy the reverse DNS name and create one
+
+`io.github.eiffel-community.<Category>.<Eiffel type>`
+
+Where:
+
+- `io.github.eiffel-community` is the reverse DNS of our website
+- `<Category>` is taken from [Event Categories](event-categories.md) but lowercase
```
Comment (magnusbaeck) [inline eiffel-syntax-and-usage/cloud-events.md:96]: Why? That requires more parties to be aware of that mapping table.

**Review by magnusbaeck — COMMENTED** [review]
Yes, I definitely think we should use the recommended reverse DNS pattern for the type. Too bad we don't have a real domain for Eiffel.

File: eiffel-syntax-and-usage/cloud-events.md (hunk 2 above)
Comment (m-linner-ericsson) [reply eiffel-syntax-and-usage/cloud-events.md:96]: Remove it

File: eiffel-syntax-and-usage/cloud-events.md (hunk 1 above)
Comment (m-linner-ericsson) [reply eiffel-syntax-and-usage/cloud-events.md:97]: Take raw event name without lowercase and hyphen

**Review by e-backmark-ericsson — APPROVED** [review]

**Review by magnusbaeck — CHANGES_REQUESTED** [review]
LGTM content-wise, but we should link to it from README.md.

**Comment by m-linner-ericsson** [comment]
> LGTM content-wise, but we should link to it from README.md.

Thanks Magnus! Fixed in [be351d1](https://github.com/eiffel-community/eiffel/pull/400/commits/be351d136d66858f240ab9a5eb0ab0266b183dad)

**Review by magnusbaeck — APPROVED** [review]

**Review by e-backmark-ericsson — APPROVED** [review]

**Pull request merged**